apiml/src/main/java/org/zowe/apiml/controller/ReactivePublicJWKController.java

```diff
@@ -0,0 +1,205 @@
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RestController;
+import org.zowe.apiml.message.core.MessageService;
+import org.zowe.apiml.security.common.token.OIDCProvider;
+import org.zowe.apiml.zaas.security.service.JwtSecurity;
+import org.zowe.apiml.zaas.security.service.token.OIDCTokenProviderJWK;
+import org.zowe.apiml.zaas.security.service.zosmf.ZosmfService;
+import reactor.core.publisher.Mono;
+
+import java.io.IOException;
+import java.io.StringWriter;
+import java.security.PublicKey;
+import java.util.Collections;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+
+import static org.zowe.apiml.zaas.controllers.AuthController.ALL_PUBLIC_KEYS_PATH;
+import static org.zowe.apiml.zaas.controllers.AuthController.CURRENT_PUBLIC_KEYS_PATH;
+import static org.zowe.apiml.zaas.controllers.AuthController.PUBLIC_KEYS_PATH;
+
+@RestController
+@RequestMapping("/gateway/api/v1/auth")
+@Slf4j
+@RequiredArgsConstructor
+public class ReactivePublicJWKController {
+
+    @Nullable private final OIDCProvider oidcProvider;
+    private final JwtSecurity jwtSecurity;
+    private final ZosmfService zosmfService;
+    private final MessageService messageService;
+
+    /**
+     * Return all public keys involved at the moment in the ZAAS as well as in zOSMF. Keys used for verification of
+     * tokens
+     *
+     * @return Map of keys composed of zOSMF and ZAAS ones
+     */
+    @GetMapping(path = ALL_PUBLIC_KEYS_PATH, produces = MediaType.APPLICATION_JSON_VALUE)
+    @Operation(summary = "Returns all public keys to verify JWT tokens validity",
+        tags = {"Security"},
+        operationId = "GetAllPublicKeysUsingGET",
+        description = "This endpoint returns all possible JWKs, which can verify sign outside the Gateway. It can contain public keys of Zowe and z/OSMF."
+    )
+    @ApiResponses(value = {
+        @ApiResponse(responseCode = "200", description = "OK",
+            content = @Content(
+                mediaType = MediaType.APPLICATION_JSON_VALUE,
+                schema = @Schema(implementation = JWKSet.class)
+            )
+        )
+    })
+    public Mono<Map<String, Object>> getAllPublicKeys() {
+        return Mono.fromSupplier(() -> {
+            List<JWK> keys;
+            if (jwtSecurity.actualJwtProducer() == JwtSecurity.JwtProducer.ZOSMF) {
+                keys = new LinkedList<>(zosmfService.getPublicKeys().getKeys());
+            } else {
+                keys = new LinkedList<>();
+            }
+            Optional<JWK> key = jwtSecurity.getJwkPublicKey();
+            key.ifPresent(keys::add);
+        if ((oidcProvider != null) && (oidcProvider instanceof OIDCTokenProviderJWK oidcTokenProviderJwk)) {
+            JWKSet oidcSet = oidcTokenProviderJwk.getJwkSet();
+            if (oidcSet != null) {
+                keys.addAll(oidcSet.getKeys());
+            }
+        }
```

Comment:
```suggestion
            if ((oidcProvider != null) && (oidcProvider instanceof OIDCTokenProviderJWK oidcTokenProviderJwk)) {
                JWKSet oidcSet = oidcTokenProviderJwk.getJwkSet();
                if (oidcSet != null) {
                    keys.addAll(oidcSet.getKeys());
                }
            }
```

Just indent, the reset is described in #4176.